### PR TITLE
Non thread dispatcher

### PR DESCRIPTION
**Pull request by xiong-gang**
1) Move cdbgang.c and cdbconn.c to dispatcher directory.
2) Remove some dead code.
3) Refactor cdbdisp_thread.c and cdbdispatchresult.c.
4) Dispatch and create gang don't using threads. Use GUC gp_connections_per_thread to control, gp_connections_per_thread equals zero means don't use threads.

### CODE REVIEW

**Comment by hlinnaka** [comment]
Wow, large PR :-). Not surprisingly, I like the general idea of using asynchronous libpq, and the removal of dead code and other clean up is obviously good. I'm going to read it through, but some high-level comments:

* I would encourage just copy-pasting more code from the threaded version to the async version, rather than trying to share code. We don't want to maintain both for more than a release or two, so there's not much to gain by trying to modularize functions to server both masters. And the reason we're keeping the threaded version at all is that we want to be conservative and allow falling back to what we know to work. It'd be good to avoid changing the old code, because any changes increase the risk of bugs and behaviour changes, and we want the fallback to behave as much as it used to as possible. So let's do what makes most sense for the non-threaded code, and copy-paste what we need, leaving the old code untouched as much as possible.

* Needs a lot more comments. Especially the new interfaces need comments explaining how they work, what the arguments are, etc.

* Since this is new code, would be good to run pgindent on it to get the formatting correct, right from the start.

File: src/backend/cdb/cdbutil.c [hunk 1]
```diff
@@ -1331,3 +1331,59 @@ bool isSockAlive(int sock)
 
 	return true;
 }
+
+/*
+ * Helper functions
```
Comment (hlinnaka) [inline src/backend/cdb/cdbutil.c]: Not a very helpful comment.

I'd suggest moving this to a separate file under cdb/dispatcher, or somewhere else. Basically, I don't want anyone else to start using gp_pthread_create() and introduce new threading in the backend. Let's put it somewhere closer to the threaded dispatcher code, and make it clear that it's it only exists to server the legacy threaded dispatcher.

File: src/backend/cdb/cdbvars.c [hunk 2]
```diff
@@ -655,6 +656,8 @@ assign_gp_connections_per_thread(int newval, bool doit, GucSource source __attri
 		if (newval < 0)
 			return false;
 
+		cdbdisp_useThread(newval > 0);
```
Comment (hlinnaka) [inline src/backend/cdb/cdbvars.c]: gp_connections_per_thread is currently marked as PGC_SUSET, which means that a superuser can change it on the fly, in an existing session. Does the code cope with that?

File: src/backend/cdb/dispatcher/cdbconn.c [hunk 3]
```diff
@@ -366,6 +367,12 @@ void cdbconn_doConnect(SegmentDatabaseDescriptor *segdbDesc, const char *gpqeid,
 
 	Assert(nkeywords < MAX_KEYWORDS);
 
+	if (!wait)
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbconn.c]: It'd probably make sense to split this function into two: one to construct the keywords/values arrays, and another to do the waiting. The non-threaded version wouldn't need the waiting part, it could just call PQconnectStartParams directly.

Note that "connect_timeout" option is a no-op with PQconnectStartParams(). I guess it doesn't do any harm either, but it'd seem nicer to not pass it to libpq if we know it won't do anything. At least it deserves a comment explaining that it's up to the caller to implement the timeout with !wait.

File: src/backend/cdb/dispatcher/cdbconn.c [hunk 4]
```diff
@@ -495,7 +514,8 @@ bool cdbconn_discardResults(SegmentDatabaseDescriptor *segdbDesc,
 /* Return if it's a bad connection */
 bool cdbconn_isBadConnection(SegmentDatabaseDescriptor *segdbDesc)
 {
-	return PQstatus(segdbDesc->conn) == CONNECTION_BAD;
+	return (PQsocket(segdbDesc->conn) < 0 ||
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbconn.c]: Is this change really needed?

File: src/backend/cdb/dispatcher/cdbdisp_non_thread.c [hunk 5]
```diff
@@ -0,0 +1,757 @@
+
+/*-------------------------------------------------------------------------
+ *
+ * cdbdisp_non_thread.c
+ *	  Functions for non-threaded implementation of dispatching
+ *	  commands to QExecutors.
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdisp_non_thread.c]: A high-level description of how this works would be in order. In particular, make sure to mention that this uses libpq's asynchronous APIs, instead of threads.

It might be good to rename this file and the corresponding functions etc. to not be "non-threaded", but "async libpq" or something. I.e. describe what it is, not what it isn't. A few years down the line when the threaded version is removed altogether, it's going to look bizarre that it's called "non-threaded".

File: src/backend/cdb/dispatcher/cdbdisp_non_thread.c [hunk 6]
```diff
@@ -0,0 +1,757 @@
...
+#include "cdb/cdbdispatchresult.h"
+#include "cdb/cdbfts.h"
+#include "cdb/cdbgang.h"
+#include "cdb/cdbvars.h"
+#include "miscadmin.h"
+
+/*
+ * Keeps state of all the dispatch command threads.
+ */
+typedef struct CdbDispatchCmdThreads
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdisp_non_thread.c]: this is the non-threaded version, no threads here

File: src/backend/cdb/dispatcher/cdbdisp_non_thread.c [hunk 7]
```diff
@@ -0,0 +1,757 @@
...
+	pParms->dispatchCount = 0;
+	pParms->waitMode = DISPATCH_WAIT_NONE;
+	pParms->query_text = queryText;
+	pParms->query_text_len = len;
+
+	return (void*)pParms;
+}
+
+
+static void
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdisp_non_thread.c]: please add a comment explaining what this function does.

File: src/backend/cdb/dispatcher/cdbconn.c (hunk 4 above)
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbconn.c]: I think it would actually be more clear to just call PQstatus() directly, and remove this function altogether.

File: src/backend/cdb/dispatcher/cdbdisp_non_thread.c [hunk 8]
```diff
@@ -0,0 +1,757 @@
...
+			if (PQisBusy(dispatchResult->segdbDesc->conn))
+				elog(LOG, "We thought we were done, because finished==true, but libpq says we are still busy");
+		}
+		else
+			ELOG_GANG_DEBUG("processResults says we have more to do with %d of %d (%s)",
+								 i + 1, pParms->dispatchCount, segdbDesc->whoami);
+	}
+}
+
+static void
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdisp_non_thread.c]: What does this function do? comment, please

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 9]
```diff
@@ -293,28 +280,311 @@ allocateWriterGang()
...
+}
+
+/*
+ * Creates a new gang by logging on a session to each segDB involved.
+ *
+ * call this function in GangContext memory context.
+ * elog ERROR or return a non-NULL gang.
+ */
+static Gang*
+createGang_non_thread(GangType type, int gang_id, int size, int content)
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbgang.c]: How about moving this to cdbdisp_non_threaded.c? It'd be nice to have all the code that's specific to the threaded or non-threaded implementations to be in the separate cdbdisp_[non_]threaded.c file.

**Comment by hlinnaka** [comment]
+1 on the first three commits ("Move cdbgang.c and cdbconn.c to dispatcher directory", "Don't allow gp_connections_per_thread to be 0", "Remove dead code").

The commit message of the second commit ("Don't allow gp_connections_per_thread to be 0") deserves more explanation. gp_connections_per_thread=0 used to mean: don't use threads, establish the connections and dispatch the query in the main thread. That option is now gone. I think that's acceptable and a good thing, because we're replacing it with the async libpq implementation. But if you just look at that single commit, you'll wonder why it was removed. So, it deserves a quick mention of what gp_connections_per_thread=0 used to mean, why we're removing it, and why we think that's OK.

File: src/backend/cdb/dispatcher/cdbdisp_thread.c [hunk 10]
```diff
@@ -775,42 +669,40 @@ thread_DispatchCommand(void *arg)
  * NOTE: since this is called via a thread, the same rules apply as to
  *		 thread_DispatchCommand absolutely no elog'ing.
  */
-static void
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdisp_thread.c]: comment on what the return value means

File: src/backend/cdb/dispatcher/cdbdisp_thread.c [hunk 11]
```diff
@@ -819,35 +711,17 @@ dispatchCommand(CdbDispatchResult * dispatchResult,
...
 /*
  * Helper function to thread_DispatchCommand that handles errors that occur
  * during the poll() call.
- *
- * NOTE: since this is called via a thread, the same rules apply as to
- *		 thread_DispatchCommand absolutely no elog'ing.
- *		 The cleanup of the connections will be performed by handlePollTimeout().
  */
 static void
-handlePollError(DispatchCommandParms * pParms, int db_count, int sock_errno)
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdisp_thread.c]: Isn't this still called in the helper thread? Why remove the NOTE comment?

**Comment by hlinnaka** [comment]
+1 on these parts of the fourth commit ("Refactor cdbdisp_thread.c and cdbdispatchresult.c "):

* the addOptions() -> makeOptions() refactoring
* removal of the 'verbose' arguments (they were dead code, all callers passed false)

I haven't fully grokked the rest yet.

I'd suggest that you proceed and push those non-controversial cleanups and refactorings  from the first three commits plus the above parts of the fourth commit, and rebase or resubmit the PR of the rest after that. Will make review easier to get those parts out of the way.

File: src/backend/cdb/cdbutil.c (hunk 1 above)
Comment (xiong-gang) [inline src/backend/cdb/cdbutil.c]: Right now ftsprobe.c is using this as well.
I think we can keep it here for now, and move it to ftsprobe.c when we can finally drop thread in backend process.

Comment (hlinnaka) [inline src/backend/cdb/cdbutil.c]: On 17/06/16 11:40, xiong-gang wrote:
>> @@ -1331,3 +1331,59 @@ bool isSockAlive(int sock)
>>
>>   	return true;
>>   }
>> +
>> +/*
>> + * Helper functions
>
> Right now ftsprobe.c is using this as well.
> I think we can keep it here for now, and move it to ftsprobe.c when we can finally drop thread in backend process.

Oh! We should probably refactor ftsprobe.c to also use asynchronous mode 
instead. It's almost there already, it already sets the socket to 
non-blocking mode after connecting and uses poll() to wait. But that's 
another story..

- Heikki

File: src/backend/cdb/cdbvars.c (hunk 2 above)
Comment (xiong-gang) [inline src/backend/cdb/cdbvars.c]: Good point.
Normally it should be ok, when the session is available to run the SET command, all threads are already finished.

There's a problem when considering named portal like this:
begin;
set gp_connections_per_thread = 10;
declare c2 cursor for select * from t;
set gp_connections_per_thread = 0;
end;

The dispatcher result of named portal is checked at the end of the transaction, but it's already changed from thread to non-thread.
Maybe we can resolve it by disable "set gp_connections_per_thread" in transaction.
Or we involve another GUC of PGC_BACKEND.
How do you think?

Comment (hlinnaka) [inline src/backend/cdb/cdbvars.c]: On 17/06/16 12:13, xiong-gang wrote:
>> @@ -655,6 +656,8 @@ assign_gp_connections_per_thread(int newval, bool doit, GucSource source __attri
>>   		if (newval < 0)
>>   			return false;
>>
>> +		cdbdisp_useThread(newval > 0);
>
> Good point.
> Normally it should be ok, when the session is available to run the SET command, all threads are already finished.
>
> There's a problem when considering named portal like this:
> begin;
> set gp_connections_per_thread = 10;
> declare c2 cursor for select * from t;
> set gp_connections_per_thread = 0;
> end;
>
> The dispatcher result of named portal is checked at the end of the transaction, but it's already changed from thread to non-thread.
> Maybe we can resolve it by disable "set gp_connections_per_thread" in transaction.
> Or we involve another GUC of PGC_BACKEND.
> How do you think?

Yeah, I think it would be acceptable to just make it PGC_BACKEND. We're 
only keeping the GUC around as an escape hatch if some serious issues 
come up with the new async libpq implementation, and it will hopefully 
go away completely in a year or two.

- Heikki

File: src/backend/cdb/dispatcher/cdbconn.c (hunk 3 above)
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbconn.c]: I tried to extract the construct keywords/values part to a new function, but didn't come out a neat result:-(
How about i add a non-thread version of this function and let them have some redundant code?

File: src/backend/cdb/dispatcher/cdbconn.c (hunk 4 above)
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbconn.c]: I refactor this function out of two concern:
1. Some code check PQsocket() and some check PQstatus(), so i merged them together.
2. I am trying to move all libpq call to cdbconn.c so i can mock this file to do unit test.
to do the test.

File: src/backend/cdb/dispatcher/cdbconn.c (hunk 3 above)
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbconn.c]: On 17/06/16 12:21, xiong-gang wrote:
>> @@ -366,6 +367,12 @@ void cdbconn_doConnect(SegmentDatabaseDescriptor *segdbDesc, const char *gpqeid,
>>
>>   	Assert(nkeywords < MAX_KEYWORDS);
>>
>> +	if (!wait)
>
> I tried to extract the construct keywords/values part to a new function, but didn't come out a neat result:-(
> How about i add a non-thread version of this function and let them have some redundant code?

Sounds good. As I opined earlier, don't be shy of duplicating code. 
Better to keep the current implementation untouched as much as possible, 
and copy-paste what we need to the new implementation.

- Heikki

File: src/backend/cdb/dispatcher/cdbconn.c (hunk 4 above)
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbconn.c]: On 17/06/16 12:26, xiong-gang wrote:
>> @@ -495,7 +514,8 @@ bool cdbconn_discardResults(SegmentDatabaseDescriptor *segdbDesc,
>>   /* Return if it's a bad connection */
>>   bool cdbconn_isBadConnection(SegmentDatabaseDescriptor *segdbDesc)
>>   {
>> -	return PQstatus(segdbDesc->conn) == CONNECTION_BAD;
>> +	return (PQsocket(segdbDesc->conn) < 0 ||
>
> I refactor this function out of two concern:
> 1. Some code check PQsocket() and some check PQstatus(), so i merged them together.
> 2. I am trying to move all libpq call to cdbconn.c so i can mock this file to do unit test.
> to do the test.

I don't like moving all libpq calls to cdbconn.c. It just adds a new 
layer for little purpose.

How about mocking libpq directly?

- Heikki

Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbconn.c]: There are too many libpq calls, and some of them need to be put together to mean something.
Compiling the libpq files also makes it's harder.
To be frankly, i don't like to wrap every libpq functions neither.
I will think about it when writing the tests.

File: src/backend/cdb/dispatcher/cdbdisp_non_thread.c (hunk 5 above)
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp_non_thread.c]: Good suggestion. Will do

File: src/backend/cdb/dispatcher/cdbdisp_non_thread.c (hunk 7 above)
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp_non_thread.c]: will do

File: src/backend/cdb/dispatcher/cdbdisp_non_thread.c (hunk 6 above)
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp_non_thread.c]: forgot to rename it when copy&past

File: src/backend/cdb/dispatcher/cdbgang.c (hunk 9 above)
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbgang.c]: I tried to create cdbgang_thread.c and cdbgang_non_thread.c files to put different implementation of createGang. But unlike dispatch, createGang is much simpler(1 or 2 functions), and they share all most all helper functions.

Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbgang.c]: On 17/06/16 16:57, xiong-gang wrote:
>> +			timeout = gp_segment_connect_timeout * 1000 - diff_us / 1000;
>> +	}
>> +	else
>> +		timeout = -1;
>> +
>> +	return timeout;
>> +}
>> +
>> +/*
>> + * Creates a new gang by logging on a session to each segDB involved.
>> + *
>> + * call this function in GangContext memory context.
>> + * elog ERROR or return a non-NULL gang.
>> + */
>> +static Gang*
>> +createGang_non_thread(GangType type, int gang_id, int size, int content)
>
> I tried to create cdbgang_thread.c and cdbgang_non_thread.c files to put different implementation of createGang. But unlike dispatch, createGang is much simpler(1 or 2 functions), and they share all most all helper functions.

I think you could just put them in cdbdisp_thread.c and 
cdbdisp_non_thread.c. I don't see much need to separate the 
implementation of gang creation code and implementation of the 
dispatching. The two always go together in practice, and as you said, 
the gang creation is not much code, so it wouldn't bloat 
cdbdisp_[non_]thread too much.

- Heikki

**Comment by xiong-gang** [comment]
@hlinnaka 
Thanks for the reviewing and good suggestions.
I will merge part of the commits as you suggested.

**Comment by xiong-gang** [comment]
@hlinnaka 
I pushed the first 3 commits and did rebase.
Could you please review it?

Thanks.

File: src/backend/cdb/dispatcher/cdbdisp.c [hunk 12]
```diff
@@ -391,3 +391,43 @@ cdbdisp_clearGangActiveFlag(CdbDispatcherState * ds)
 		ds->primaryResults->writer_gang->dispatcherActive = false;
 	}
 }
+
+void
+CollectQEWriterTransactionInformation(SegmentDatabaseDescriptor * segdbDesc,
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdisp.c]: This really belongs in libpq itself. cdbdisp.c should not be peeking into fields in PGconn directly.

**Comment by hlinnaka** [comment]
Thanks for the rebasing.

The now 2nd patch in the series, "Change the retry logic of gang creation", seems unrelated to the rest of the changes. Can you post that as a separate PR, so that we can discuss that separately, please?

File: src/backend/cdb/dispatcher/cdbdispatchresult.c [hunk 13]
```diff
@@ -445,18 +421,17 @@ cdbdisp_debugDispatchResult(CdbDispatchResult * dispatchResult,
...
-
-		/*
-		 * QE error or libpq error
-		 */
 		else
 		{
 			char *sqlstate = PQresultErrorField(pgresult, PG_DIAG_SQLSTATE);
 			char *pri = PQresultErrorField(pgresult, PG_DIAG_MESSAGE_PRIMARY);
 			char *dtl = PQresultErrorField(pgresult, PG_DIAG_MESSAGE_DETAIL);
+			char *sourceFile = PQresultErrorField(pgresult, PG_DIAG_SOURCE_FILE);
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdispatchresult.c]: It's theoretically possible for sourceFile or sourceLine to be NULL. The ereport() below should be prepared for that (yeah, the old code wasn't prepared for that either, but let's fix it while we're at it)

**Comment by xiong-gang** [comment]
I create a separate PR:
https://github.com/greenplum-db/gpdb/pull/856

File: src/backend/cdb/dispatcher/cdbdisp_query.c [hunk 14]
```diff
@@ -768,20 +769,14 @@ CdbDispatchUtilityStatement_NoTwoPhase(struct Node *stmt,
  *
  * Also, we free the DispatchCommandQueryParms memory.
  */
-static void
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdisp_query.c]: The comment above is outdated.

File: src/backend/cdb/dispatcher/cdbdisp.c (hunk 12 above)
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp.c]: OK, i will move it to libpq or cdbconn.c
This seems like dead code by the way.

**Comment by hlinnaka** [comment]
Ok, this is still quite a beast to review. Let's try to split it further, into more palatable pieces.

The change to remove maxCapacity argument from cdbdisp_makeDispatchResults looks OK to me. Can you separate that into a commit that does just that, please?

File: src/backend/cdb/dispatcher/cdbdisp.c (hunk 12 above)
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbdisp.c]: On 20/06/16 12:40, xiong-gang wrote:
>> @@ -391,3 +391,43 @@ cdbdisp_clearGangActiveFlag(CdbDispatcherState * ds)
>>   		ds->primaryResults->writer_gang->dispatcherActive = false;
>>   	}
>>   }
>> +
>> +void
>> +CollectQEWriterTransactionInformation(SegmentDatabaseDescriptor * segdbDesc,
>
> OK, i will move it to libpq or cdbconn.c
> This seems like dead code by the way.

Oh, ok, let's remove it then :-)

- Heikki

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 15]
```diff
@@ -293,28 +280,311 @@ allocateWriterGang()
...
+				ereport(LOG, (errcode(ERRCODE_GP_INTERNAL_ERROR),
+							  errmsg("Master unable to connect to %s with options %s: %s",
+								segdbDesc->whoami,
+								options,
+								PQerrorMessage(segdbDesc->conn))));
+
+				ereport(ERROR, (errcode(ERRCODE_GP_INTERCONNECTION_ERROR),
+								errmsg("failed to acquire resources on one or more segments")));
+			}
+			segdbDesc->pollStatus = PGRES_POLLING_WRITING;
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbgang.c]: After calling PQconnectStartParams(), you are supposed to call PQconnectPoll() next to figure out whether you need to wait for the socket to become writeable or readable.

I think you should refactor the loop below so that it calls PQconnectPoll() on each connection first, not last. That would allow you to also get rid of the pollStatus field in SegmentDatabaseDescriptor. That really doesn't belong there.

**Comment by hlinnaka** [comment]
+1 on moving AssertSliceTableIsValid(). That's another independent change that I think you could push separately.

I think you've gotten plenty of comments and TODOs from this review now. I'll let you deal with them, and spend a while refactoring, cleaning up, rebasing etc.

**Comment by xiong-gang** [comment]
I merged the changes @hlinnaka which thinks is good.
I splited the rest of the changes into smaller commits. I will close this PR and open a new one.
